### DISCUSSION TITLE
Connection Pool close connections when max_overflow is greater than 0 (asyncpg)

### DISCUSSION THREAD

**Discussion by augustomarinho** [sqlalchemy/sqlalchemy#11707] [Usage Questions]
Hi, guys. I have a problem using SQLAlchemy + AsyncPG when I configure the _max_overflow_ greater than 0 and I found this issue https://github.com/sqlalchemy/sqlalchemy/issues/7059.

I saw the trouble in asyncpg, but it wasn´t clear what I need to configure to solve this issue.

I'm using SQLAlchemy 2.0.30 and asyncpg 0.29.0. The asyncpg fix was released at v0.25.0 (https://github.com/MagicStack/asyncpg/issues/830), but I continue to experiment with this scenario in Linux (Docker container with image python:3.11-slim)

### Only for example
I used the code below:
```
db = create_async_engine("postgresql+asyncpg://USER:PASSWORD@HOST:PORT/dbName", isolation_level="AUTOCOMMIT", pool_size="10", max_overflow="20", pool_recycle="3600", pool_timeout=30, echo_pool="debug", echo=True)
```

```
class UrlRepository(UrlBaseRepository):
    def __init__(self, db: AsyncEngine):
        self.db = db
        
    async def get_url_by_short_code(self, short_code: str):
        async with self.db.connect() as conn:
            result = await conn.execute(<<QUERY>>)
            url_data = result.fetchone()
            
        return url_data
```

Here is a couple of lines of log messages. 
```
2024-08-07 01:40:21,199 INFO sqlalchemy.engine.Engine ROLLBACK using DBAPI connection.rollback(), DBAPI should ignore due to autocommit mode
2024-08-07 01:40:21,199 DEBUG sqlalchemy.pool.impl.AsyncAdaptedQueuePool Connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7eed7d3731f0>> being returned to pool
2024-08-07 01:40:21,199 DEBUG sqlalchemy.pool.impl.AsyncAdaptedQueuePool Connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7eed7d3731f0>> reset, transaction already reset
2024-08-07 01:40:21,200 DEBUG sqlalchemy.pool.impl.AsyncAdaptedQueuePool Closing connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7eed7d3731f0>>
```

### Conclusion
This scenario happens when the application has a burst and the app needs to use more connections than the default (max_overflow).

When the application needs to use less than 10 connections, this isn't seen. But there are many idle connections detected (another scenario).

**Comment by CaselIT** [comment]
Hi,

It's not overly clear what your issue is. That log is expected when overflowing the pool.

The linked issue refers a memory leak, can you share some details about what happens in your case?

**Reply by augustomarinho** [reply]
I have observed some disconnection logs in Postgres. To simulate a burst moment in the application, I configured **pool_size=10** and **max_overflow=20** to simulate a load test. I used 30 users simultaneously for 1 minute (I ran multiple scenarios; this is just one example).

In this scenario, when the application needs to use more than 10 connections and gets "extra" connections from the pool, these connections are returned to the pool and then closed. I would like to understand if this scenario is expected behavior. During a burst, should the connection pool close connections?

I have the impression that there is a direct correlation between HTTP requests and active database connections. This makes sense, but I'm unsure if the connection pool should close "extra" connections instead of returning them to the pool. Is there a way to improve this behavior, such as using a parameter to set a "time to hold" the connections (extra) before closing them?

I haven't tested using pgBouncer in front of the database yet to see if it improves the behavior. However, I think there might be something unclear to me about the expected behavior of the connection pool when using overflowed connections.

**Reply by CaselIT** [reply]
> In this scenario, when the application needs to use more than 10 connections and gets "extra" connections from the pool, these connections are returned to the pool and then closed. I would like to understand if this scenario is expected behavior. During a burst, should the connection pool close connections?

yes, that's what the max_overflow parameter controls: https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.QueuePool.params.max_overflow

> I have the impression that there is a direct correlation between HTTP requests and active database connections. This makes sense, but I'm unsure if the connection pool should close "extra" connections instead of returning them to the pool. Is there a way to improve this behavior, such as using a parameter to set a "time to hold" the connections (extra) before closing them?

the queue pool holds at most `pool_size` connections. all extra connections will be single use.
There is currently no parameter to achieve what you want using the queue pool, but you could probably approximate it using https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.QueuePool.params.use_lifo and a server side timeout, so that you can configure the server to close inactive connection after a while. In this case you should also set up https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.Pool.params.recycle or pre-ping. use lifo may also not be necessary depending on your usage.

from your description I don't think this is related to the linked issue

**Reply by augustomarinho** [reply]
Thanks for you reply @CaselIT . The most valuable information for me is about "... extra connections will be single use." In this case, I'll execute a couple of tests using another approach.

Related to the `pool_use_lifo`, `pool_pre_ping`, and `pool_recycle`, I validated them and I didn't see any improvement for this scenario (in my case).

Related to idle connections, I'll check the server-side configurations.

Thank you once again for your clarifications. They were very helpful to me.

**Reply by CaselIT** [reply]
> Related to the pool_use_lifo, pool_pre_ping, and pool_recycle, I validated them and I didn't see any improvement for this scenario (in my case).

my suggestion here was to avoid the overflow and make the pool size bigger. This avoids single use connections and you let the server close idle connection after a while
